README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# The Johnson Quantile-Parameterised Distribution

<!-- badges: start -->

<!-- badges: end -->

The goal of `rjqpd` is to implement the Johnson Quantile-Parameterised
distribution system in R. It provides the density function, cumulative
distribution function, quantile function and random sampling function in
their usual form:

  - `djqpd` - calculates the density function (PDF)
  - `pjqpd` - calculates the cumulative distribution function (CDF)
  - `qjqpd` - calculates the quantile function (QF)
  - `rjqpd` - generates random samples

## Background

A common problem in data analysis is encountering continuous
uncertainties that are not easily characterised by well known
parametric, continuous probability distributions. Often the underlying
process that generated the data is unknown or is inconvenient to express
in closed form. In these cases one typically has to resort to curve
fitting a distribution to a set of probability-quantile pairs without
the guarantee that the best-fit distribution will pass through the
assessed points exactly.

The Johnson Quantile-Parameterised Distribution (J-QPD) system is a
novel approach to this problem first developed in a
[paper](https://pubsonline.informs.org/doi/10.1287/deca.2016.0343) by
Hadlock and Bickel (2017) in Decision Analysis Vol. 14, No. 1.

Their approach is to extend the [Johnson Distribution
System](https://en.wikipedia.org/wiki/Johnson's_SU-distribution) in a
way such that the distributions constructed honour exactly a symmetric
percentile triplet of quantile assessments (e.g. 15th, 50th, 85th) on a
bounded or semi-bounded domain. By parameterising the distribution in
this way the need for curve fitting can be eliminated.

## Installation

The `rjqpd` package is not yet available on CRAN.

You can install the latest development version of `rjqpd` using:

``` r
devtools::install_github("bobbyingram/rjqpd")
```

## Basic Usage

The base function of the package is the `jqpd` class constructor:

``` r
jqpd()
```

This function takes the convenient symmetric percentile triplet quantile
parameterisation and computes a set of constants required to calculate
the distribution functions.

## Basic Example

This is a basic example to show how to calculate the parameters of a
bounded Johnson Quantile-Parameterised distribution from the quantiles
of a symmetric percentile triplet (SPT).

``` r
library(rjqpd)

# alpha = 0.1 indicates a symmetric percentile triplet (SPT) of (0.1, 0.5, 0.9)
alpha <- 0.1
# we parameterise using some quantile values corresponding to the SPT above.
quantiles <- c(0.32, 0.40, 0.60)
params <- jqpd(quantiles, lower = 0, upper = 1, alpha = alpha)

# The CDF should return the (SPT) at the quantile values:
pjqpd(quantiles, params)
#> [1] 0.1 0.5 0.9
```

### Summary Plot

The `plot_jqpd` function produces a simple summary plot showing the
density, cumulative distribution function, quantile function and 1000
random samples:

``` r
plot_jqpd(params)
```

<img src="man/figures/README-summary_plot-1.png" width="100%" />

### Summary Statistics

We also provide some basic summary statistics functions:

  - `jqpd_mean` - calculates the mean
  - `jqpd_var` - calculates the variance
  - `jqpd_sd` - calculates the standard deviation
  - `jqpd_skewness` - calculates the skewness
  - `jqpd_kurtosis` - calculates the kurtosis

<!-- end list -->

``` r
jqpd_mean(params)
#> [1] 0.4350185
```

``` r
jqpd_var(params)
#> [1] 0.01604665
```

``` r
jqpd_sd(params)
#> [1] 0.1266754
```

``` r
jqpd_skewness(params)
#> [1] 1.69221
```

``` r
jqpd_kurtosis(params)
#> [1] 6.464129
```


NEWS.md
# rjqpd 0.2.3

* Restore user graphical parameter settings on exit of `plot_jqpd`.
* Documented return values of moment functions.

# rjpqd 0.2.2

* Fixed DOI in DESCRIPTION file.

# rjqpd 0.2.1

* Fixed bug in `plot_jqpd` for semi-bounded distributions.

# rjqpd 0.2.0

* Added `plot_jqpd` function
* Added `jqpd_mean`, `jqpd_var`, `jqpd_sd`, `jqpd_skewness` and `jqpd_kurtosis`
functions.

# rjqpd 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(rjqpd)

test_check("rjqpd")


tests/testthat/test-djqpd.R
context("density function")

test_models <- list(
  jqpd(c(0.32, 0.4, 0.6), lower = 0, upper = 1, alpha = .1),
  jqpd(c(0.32, 0.4, 0.6), lower = 0, upper = 1, alpha = .3),
  jqpd(c(2, 4, 6), lower = 0, upper = Inf, alpha = .1),
  jqpd(c(2, 4, 6), lower = 0, upper = Inf, alpha = .3)
)

test_that("outside bounds density is NaN", {
   for (params in test_models) {
     expect_equal(djqpd(params$lower - 1, params), NaN)
     if(!is.infinite(params$upper)){
       expect_equal(djqpd(params$upper + 1, params), NaN)
     }
   }
})

test_that("at bounds density is 0", {
  for (params in test_models) {
    expect_equal(djqpd(params$lower, params), 0)
    expect_equal(djqpd(params$upper, params), 0)
  }
})

test_that("within bounds density is non-zero", {
  for (params in test_models) {
    density <- djqpd(params$x, params)
    expect_true(all(density > 0))
  }
})



tests/testthat/test-pjqpd.R
context("cumulative distribution function")

test_models <- list(
  jqpd(c(0.32, 0.4, 0.6), lower = 0, upper = 1, alpha = .1),
  jqpd(c(0.32, 0.4, 0.6), lower = 0, upper = 1, alpha = .3),
  jqpd(c(2, 4, 6), lower = 0, upper = Inf, alpha = .1),
  jqpd(c(2, 4, 6), lower = 0, upper = Inf, alpha = .3)
)

test_that("lower bound returns probability zero", {
  for (params in test_models) {
    expect_equal(pjqpd(params$lower, params), 0)
  }
})

test_that("upper bound returns probability one", {
  for (params in test_models) {
    expect_equal(pjqpd(params$upper, params), 1)
  }
})

test_that("recovers inputs", {
  for (params in test_models) {
    p <- pjqpd(params$x, params)
    expect_equal(p[1], params$alpha)
    expect_equal(p[2], 0.5)
    expect_equal(p[3], 1-params$alpha)
  }
})


R/plot.R
#' Plots the density, cumulative distribution function, quantile function and a
#' set of 1000 random samples for a Johnson Quantile-Parameterised Distribution.
#'
#' @param params jqpd object created using \code{jqpd()}
#' @return no return value, called for side effects only
#' @export
plot_jqpd <- function(params){
  if(!is_jqpd_obj(params)){
    stop("Input 'params' must have class 'jqpd'.")
  }
  lower <- params$lower
  upper <- params$upper
  if (is.infinite(upper)){
    upper <- qjqpd(0.999, params)
  }

  x <- seq(lower, upper, length.out = 1001)
  p <- seq(0.001, 0.999, length.out = 1001)
  density <- djqpd(x, params)
  cdf <- pjqpd(x, params)
  quantile <- qjqpd(p, params)
  samples <- rjqpd(1000, params)

  old <- graphics::par(mfrow = c(2, 2), xaxs = "i", yaxs = "i")
  on.exit(graphics::par(old), add = TRUE)

  plot(x, density,
       type = "l", main = "PDF", xlab = "x", ylab = "Density")
  plot(x, cdf, type = "l", main = "CDF", xlab = "x", ylab = "P(X<=x)")
  plot(p, quantile, type = "l", main = "Quantile Function", xlab = "p",
       ylab = "P(X<=x)=p")
  plot(samples, type = "p", main = "Random Sample", xlab = "sample",
       ylab = "X")
}


R/djqpd.R
#' Density function of Johnson Quantile-Parameterised Distribution.
#'
#' @param x vector of quantiles
#' @param params jqpd object created using \code{jqpd()}
#' @return A numeric vector of density values corresponding to the x quantile
#'   vector
#'
#' @export
#'
#' @examples
#' x <- c(0.32, 0.40, 0.60)
#' params <- jqpd(x, lower = 0, upper = 1, alpha = 0.1)
#' iles <- seq(0.01, 0.99, 0.01)
#' density <- djqpd(x = iles, params)
djqpd <- function(x, params){
  UseMethod("djqpd", params)
}

#' @export
djqpd.default <- function(x, params){
  print("'params' object must have class 'jqbd'.")
}

#' @export
djqpd.jqpd <- function(x, params){
  d <- numeric(length(x))
  i <- x > params$lower & x < params$upper

  # Outside of the bounds the density is undefined.
  d[!i] <- NaN
  # At the boundaries the density is zero.
  d[x == params$lower] <- 0
  d[x == params$upper] <- 0

  # Only calculate the density inside the bounds.
  if (length(x[i]) <= 0){
    return(d)
  }
  if(is_bounded(params)){
    d[i] <- density_bounded(x[i], params)
  } else{
    d[i] <- density_semibounded(x[i], params)
  }
  d
}

density_bounded <- function(x, params){
  eta <- params$eta
  lambda <- params$lambda
  delta <- params$delta
  c <- params$c
  l <- params$lower
  u <- params$upper
  n <- params$n
  k <- params$k

  z0 <- stats::qnorm((x-l)/(u-l))
  z1 <- stats::dnorm(-n*c + (1/delta)*asinh((1/lambda)*(-eta+z0)))
  z2 <- delta*(u-l)*sqrt(lambda^2 + (-eta+z0)^2)*stats::dnorm(z0)
  z1 / z2
}

density_semibounded <- function(x, params){
  eta <- params$eta
  lambda <- params$lambda
  delta <- params$delta
  c <- params$c
  l <- params$lower
  u <- params$upper
  n <- params$n
  k <- params$k

  L <- log((x-l)/eta)
  z1 <- 1 / (lambda*delta*(x-l))
  z2 <- k - (n*c*delta*(L / sqrt(lambda^2 + L^2)))
  z3 <- stats::dnorm((1/(lambda*delta))*(k*L-n*c*delta*sqrt(lambda^2 + L^2)))

  z1 * z2 * z3
}


R/pjqpd.R
#' Cumulative distribution function of Johnson Quantile-Parameterised
#' Distribution.
#'
#' @param x vector of quantiles
#' @param params jqpd object created using \code{jqpd()}
#' @return A numeric vector of probabilities corresponding to the x quantiles
#'   vector
#'
#' @export
#'
#' @examples
#' x <- c(0.32, 0.40, 0.60)
#' params <- jqpd(x, lower = 0, upper = 1, alpha = 0.1)
#' iles <- seq(0.01, 0.99, 0.01)
#' probs <- pjqpd(x = iles, params)
pjqpd <- function(x, params){
  UseMethod("pjqpd", params)
}

#' @export
pjqpd.default <- function(x, params){
  print("'params' object must have class 'jqbd'.")
}

#' @export
pjqpd.jqpd <- function(x, params){
  p <- numeric(length(x))
  i <- x > params$lower & x < params$upper
  # Outside of the bounds the CDF is undefined.
  p[!i] <- NaN
  # At the lower boundary the CDF has value 0.
  p[x == params$lower] <- 0
  # At the upper boundary the CDF has value 1.
  p[x == params$upper] <- 1

  if (length(x[i]) <= 0){
    return(p)
  }
  # Calculate the CDF at all points inside the bounds
  if(is_bounded(params)){
    p[i] <- cdf_bounded(x[i], params)
  } else {
    p[i] <- cdf_semibounded(x[i], params)
  }
  p
}

cdf_bounded <- function(x, params){
  l <- params$lower
  u <- params$upper

  eta <- params$eta
  lambda <- params$lambda
  delta <- params$delta
  n <- params$n
  c <- params$c

  stats::pnorm((1/delta)*asinh((1/lambda)*(stats::qnorm((x-l)/(u-l))-eta))-n*c)
}

cdf_semibounded <- function(x, params){
  l <- params$lower

  eta <- params$eta
  lambda <- params$lambda
  delta <- params$delta
  n <- params$n
  c <- params$c

  stats::pnorm((1/delta)*sinh(asinh((1/lambda)*log((x-l)/eta))-asinh(n*c*delta)))
}


R/qjqpd.R
#' Quantile function of Johnson Quantile-Parameterised Distribution.
#'
#' @param p vector of probabilities
#' @param params jqpd object created using \code{jqpd()}
#' @return A numeric vector of quantiles corresponding to the p probability
#'   vector
#'
#' @export
#'
#' @examples
#' x <- c(0.32, 0.40, 0.60)
#' params <- jqpd(x, lower = 0, upper = 1, alpha = 0.1)
#' probs <- seq(0.01, 0.99, 0.01)
#' quantiles <- qjqpd(p = probs, params)
qjqpd <- function(p, params){
  UseMethod("qjqpd", params)
}

#' @export
qjqpd.jqpd <- function(p, params){
  x <- numeric(length(p))
  i <- p > 0 & p < 1
  # Outside of the bounds the QF is undefined.
  x[!i] <- NaN
  # When the CDF is zero the quantile is the lower boundary.
  x[p == 0] <- params$lower
  # When the CDF is 1 the quantile is the upper boundary.
  x[p == 1] <- params$upper

  if(is_bounded(params)){
    x[i] <- quantile_bounded(p[i], params)
  } else{
    x[i] <- quantile_semibounded(p[i], params)
  }
  x
}

#' @export
qjqpd.default <- function(p, params){
  print("'params' object have class 'jqbd'.")
}

quantile_bounded <- function(p, params){
  l <- params$lower
  u <- params$upper

  eta <- params$eta
  lambda <- params$lambda
  delta <- params$delta
  n <- params$n
  c <- params$c

  l + (u-l)*stats::pnorm(eta + (lambda*sinh(delta*(stats::qnorm(p)+(n*c)))))
}

quantile_semibounded <- function(p, params){
  l <- params$lower

  eta <- params$eta
  lambda <- params$lambda
  delta <- params$delta
  n <- params$n
  c <- params$c

  l + eta*exp(lambda*sinh(asinh(delta*stats::qnorm(p))+asinh(n*c*delta)))
}


R/moments.R
moments <- function(params, n = 1, c = 0, ...){
  f <- function(x) ((x-c)^n) * djqpd(x, params)
  mu <- stats::integrate(f, params$lower, params$upper, ...)
  mu$value
}

#' Calculates the mean of a Johnson Quantile-Parameterised Distribution.
#'
#' @param params jqpd object created using \code{jqpd()}
#' @return The mean of the distribution as a length one numeric vector.
#' @export
jqpd_mean <- function(params) {
  if(!is_jqpd_obj(params)){
    stop("Input 'params' must have class 'jqpd'.")
  }
  moments(params)
}

#' Calculates the variance of a Johnson Quantile-Parameterised Distribution.
#'
#' @param params jqpd object created using \code{jqpd()}
#' @return The variance of the distribution as a length one numeric vector.
#' @export
jqpd_var <- function(params){
  if(class(params) != "jqpd"){
    stop("Input 'params' must have class 'jqpd'.")
  }
  mu <- jqpd_mean(params)
  moments(params, 2, mu)
}

#' Calculates the standard-deviation of a Johnson Quantile-Parameterised
#' Distribution.
#'
#' @param params jqpd object created using \code{jqpd()}
#' @return The standard deviation of the distribution as a length one numeric
#' vector.
#' @export
jqpd_sd <- function(params){
  sqrt(jqpd_var(params))
}

#' Calculates the skewness of a Johnson Quantile-Parameterised
#' Distribution.
#'
#' @param params jqpd object created using \code{jqpd()}
#' @return The skewness of the distribution as a length one numeric vector.
#' @export
jqpd_skewness <- function(params){
  if(class(params) != "jqpd"){
    stop("Input 'params' must have class 'jqpd'.")
  }
  mu <- jqpd_mean(params)
  sd <- jqpd_sd(params)
  moments(params, 3, mu) / sd^3
}

#' Calculates the kurtosis of a Johnson Quantile-Parameterised
#' Distribution.
#'
#' @param params jqpd object created using \code{jqpd()}
#' @return The kurtosis of the distribution as a length one numeric vector.
#' @export
jqpd_kurtosis <- function(params){
  if(class(params) != "jqpd"){
    stop("Input 'params' must have class 'jqpd'.")
  }
  mu <- jqpd_mean(params)
  var <- jqpd_var(params)
  moments(params, 4, mu) / var^2
}


R/jqpd.R
#' Calculates the parameters of the Johnson Quantile-Parameterised Distribution
#'
#' @param x a length 3 numeric vector containing the symmetric percentile
#'  triplet values used to parameterise the distribution.
#' @param lower a real number specifying the lower bound of the distribution.
#'  (default: 0)
#' @param upper a real number specifying the upper bound of the distribution. A
#'  value of Inf indicates a semi-bounded distribution. (default: Inf)
#' @param alpha a real number (between 0 and 0.5) used to describe the symmetric
#'  percentile triplet for which the quantile values provided in 'x' correspond.
#'  For instance, alpha = 0.1 (default value) indicates the percentiles used are
#'  [0.1, 0.5, 0.9].
#'
#' @return A \code{jqpd} object with elements
#' \item{x}{a length 3 numeric vector containing the symmetric percentile
#'  triplet values used to parameterise the distribution}
#' \item{alpha}{a real number (between 0 and 0.5) used to describe the symmetric
#'  percentile triplet for which the quantile values provided in 'x' correspond}
#' \item{lower}{a real number specifying the lower bound of the distribution}
#' \item{upper}{a real number specifying the upper bound of the distribution}
#' \item{c}{distribution parameter}
#' \item{n}{distribution parameter}
#' \item{eta}{distribution parameter}
#' \item{delta}{distribution parameter}
#' \item{lambda}{distribution parameter}
#' \item{k}{distribution parameter}
#' @export
#'
#' @examples
#' theta <- jqpd(c(0.32, 0.40, 0.6), 0, 1, alpha = 0.1)
jqpd <- function(x, lower = 0, upper = Inf, alpha = 0.1){

  if (!is_valid_numeric_input(x, 3)){
    stop("Input 'x' must be a length 3 numeric vector.")
  }

  if (!is_valid_bounds(lower, upper)){
    stop("Inputs 'lower' and 'upper' must be length 1 numeric vectors.")
  }

  if (!is_valid_alpha(alpha)) {
    stop("Inputs 'alpha' must be length 1 numeric vector between 0 and 0.5.")
  }

  if (!is_compatible(x, lower, upper)){
    stop("The quantiles provided are not compatible. Please ensure that:
         lower < x[1] < x[2] < x[3] < upper")
  }

  if(is.infinite(upper)){
    return(semibounded_params(x, lower, alpha))
  }
  bounded_params(x, lower, upper, alpha)
}

bounded_params <- function(x, lower, upper, alpha = 0.1){

  c <- stats::qnorm(1-alpha)
  L <- stats::qnorm((x[1] - lower)/(upper-lower))
  B <- stats::qnorm((x[2] - lower)/(upper-lower))
  H <- stats::qnorm((x[3] - lower)/(upper-lower))
  n <- sign(L + H - 2*B)

  eta <- L
  if (n == 0){
    eta <- B
  }
  if (n == -1){
    eta <- H
  }

  delta <- (1/c)*acosh((H-L)/(2*min(B-L, H-B)))
  lambda <- (H-L) / sinh(2*delta*c)

  params <- structure(list(), class = "jqpd")
  params$x <- x
  params$alpha <- alpha
  params$lower <- lower
  params$upper <- upper
  params$c <- c
  params$n <- n
  params$eta <- eta
  params$delta <- delta
  params$lambda <- lambda
  params$k <- 0
  params
}

semibounded_params <- function(x, lower, alpha = 0.1){

  c <- stats::qnorm(1-alpha)
  L <- log(x[1] - lower)
  B <- log(x[2] - lower)
  H <- log(x[3] - lower)
  n <- sign(L + H - 2*B)
  eta <- x[1] - lower
  if (n == 0){
    eta <- x[2] - lower
  }
  if (n == -1){
    eta <- x[3] - lower
  }

  delta <- (1/c)*sinh(acosh((H-L)/(2*min(B-L, H-B))))
  lambda <- (1 / (delta*c))*min(H-B, B-L)
  k <- sqrt(1+(c*delta)^2)

  params <- structure(list(), class = "jqpd")
  params$x <- x
  params$alpha <- alpha
  params$lower <- lower
  params$upper <- Inf
  params$c <- c
  params$n <- n
  params$eta <- eta
  params$delta <- delta
  params$lambda <- lambda
  params$k <- k
  params
}
